benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java

```diff
@@ -0,0 +1,436 @@
+/*
+ *
+ *  * Copyright 2017-2019 original authors
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License");
+ *  * you may not use this file except in compliance with the License.
+ *  * You may obtain a copy of the License at
+ *  *
+ *  * http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package io.micronaut.deserialization.xml;
+
+import io.micronaut.context.ApplicationContext;
+import io.micronaut.core.async.subscriber.TypedSubscriber;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.type.Argument;
+import io.micronaut.http.netty.stream.DefaultStreamedHttpRequest;
+import io.micronaut.http.server.netty.HttpContentProcessor;
+import io.micronaut.http.server.netty.NettyHttpRequest;
+import io.micronaut.http.server.netty.xml.XmlHttpContentSubscriberFactory;
+import io.micronaut.runtime.server.EmbeddedServer;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.DefaultHttpContent;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.HttpVersion;
+import io.reactivex.Flowable;
+import org.apache.groovy.util.Maps;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+public class XmlDeserializationBenchmark {
```

Comment:
@jameskleeh So here are some benchmark results. I have been running same tests against changes from current PR that  is using XmlMapper as a deserializer and changes from `xml-processing` branch that is converting xml stream to map-like structure that is later converted to object by Jackson. 

The benchmark results are not 100% fair cause `xml-processing` branch still lacks the support of collection deserialization that will require additional logic and usage of introspection functionality. So for the benchmark, I pick the xml structure that works in both branches just to get a general sense of performance degradation in case we use XmlMapper

#### That is interesting:
- Deserialization with introspected model work same or even worse results
- With map-like deserialization, on average it takes  0.015 ms/op; while with xml mapper it is 0.020. With xml-mapper it works a bit slower, but we get full XML support and JAXB would work too 

### Results from current PR (using XmlMapper)
https://gist.github.com/svishnyakov/46ea21d2d4824fcc3951dd17e0224073
Benchmark                                                                                                              Mode     Cnt   Score    Error   Units
XmlDeserializationBenchmark.streamingXmlDeserialization                                                               thrpt       4   6.531 ± 17.469  ops/ms
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization                                                   thrpt       4   7.434 ± 14.289  ops/ms
XmlDeserializationBenchmark.xmlDeserialization                                                                        thrpt       4  48.473 ±  0.968  ops/ms
XmlDeserializationBenchmark.xmlIntrospectedDeserialization                                                            thrpt       4  47.282 ±  1.216  ops/ms
XmlDeserializationBenchmark.streamingXmlDeserialization                                                              sample  408257   0.098 ±  0.001   ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.00                            sample           0.084            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.50                            sample           0.093            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.90                            sample           0.107            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.95                            sample           0.119            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.99                            sample           0.165            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.999                           sample           0.251            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.9999                          sample           1.531            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p1.00                            sample           4.612            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization                                                  sample  394288   0.101 ±  0.001   ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.00    sample           0.085            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.50    sample           0.098            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.90    sample           0.109            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.95    sample           0.123            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.99    sample           0.177            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.999   sample           0.269            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.9999  sample           1.525            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p1.00    sample           3.600            ms/op
XmlDeserializationBenchmark.xmlDeserialization                                                                       sample  983241   0.020 ±  0.001   ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.00                                              sample           0.017            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.50                                              sample           0.019            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.90                                              sample           0.022            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.95                                              sample           0.022            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.99                                              sample           0.036            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.999                                             sample           0.073            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.9999                                            sample           1.278            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p1.00                                              sample           2.744            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization                                                           sample  963280   0.021 ±  0.001   ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.00                      sample           0.018            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.50                      sample           0.020            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.90                      sample           0.022            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.95                      sample           0.023            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.99                      sample           0.035            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.999                     sample           0.069            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.9999                    sample           1.253            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p1.00                      sample           2.843            ms/op

### Results when converting to map-like structure:
https://gist.github.com/svishnyakov/00920090ac11dfa03ea93ebd6f0e320f
Benchmark                                                                                                              Mode      Cnt   Score    Error   Units
XmlDeserializationBenchmark.streamingXmlDeserialization                                                               thrpt        4  11.584 ±  0.132  ops/ms
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization                                                   thrpt        4  10.740 ±  9.396  ops/ms
XmlDeserializationBenchmark.xmlDeserialization                                                                        thrpt        4  68.143 ±  2.720  ops/ms
XmlDeserializationBenchmark.xmlIntrospectedDeserialization                                                            thrpt        4  68.557 ±  1.345  ops/ms
XmlDeserializationBenchmark.streamingXmlDeserialization                                                              sample   471401   0.085 ±  0.001   ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.00                            sample            0.071            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.50                            sample            0.080            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.90                            sample            0.090            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.95                            sample            0.101            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.99                            sample            0.198            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.999                           sample            0.327            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.9999                          sample            1.505            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p1.00                            sample            3.416            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization                                                  sample   468578   0.085 ±  0.001   ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.00    sample            0.074            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.50    sample            0.080            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.90    sample            0.089            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.95    sample            0.099            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.99    sample            0.200            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.999   sample            0.325            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.9999  sample            1.483            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p1.00    sample            2.834            ms/op
XmlDeserializationBenchmark.xmlDeserialization                                                                       sample   691697   0.015 ±  0.001   ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.00                                              sample            0.012            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.50                                              sample            0.014            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.90                                              sample            0.015            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.95                                              sample            0.016            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.99                                              sample            0.028            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.999                                             sample            0.066            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.9999                                            sample            1.238            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p1.00                                              sample            2.540            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization                                                           sample  1096537   0.018 ±  0.001   ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.00                      sample            0.016            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.50                      sample            0.017            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.90                      sample            0.018            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.95                      sample            0.023            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.99                      sample            0.032            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.999                     sample            0.076            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.9999                    sample            0.155            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p1.00                      sample            2.421            ms/op

Reply:
> With map-like deserialization, on average it takes 0.015 ms/op; while with xml mapper it is 0.020

So that is 33% slower then yeah?

Reply:
That is right. Though as I wrote it is not really fair comparison at this point due to incomplete implementation of `map-like first` solution. 
I am wondering how critical those 33% are, and if it worth to block addition of xml support because of it

Reply:
@svishnyakov No one is blocking the XML support. We've decided the way we'd like to go about doing so for good reasons (performance, less dependencies, memory consumption) and we're open to any contributing to that goal. If no one is able or willing to then we eventually will add the support. It's just a matter of priorities.

Reply:
@jameskleeh I do understand that and accept. That's being said I cannot completely agree with this, maybe because I don't believe xml support is going to become some of the top priorities in near future and even if it does, support of all xml trickiness is going to be really challenging and restrictive(due to required introspection) and probably incomplete support of all the legacy xml stuff like for SOAP,JAXB etc.

Reply:
> support of all xml trickiness is going to be really challenging and restrictive(due to required introspection) and probably incomplete support of all the legacy xml stuff like for SOAP,JAXB etc.

I don't think requiring introspection adds any challenges or restrictions. I don't know how either your proposed solution or the current direction addresses anything to do with SOAP or JAXB.

Reply:
Speaking of JAXB, it is quite easy if Jackson takes cares about xml deserialization, https://github.com/FasterXML/jackson-modules-base/tree/master/jaxb; just apply a module for xml mapper and you get jaxb support